solver/src/solver/matcha_solver.rs

```diff
@@ -0,0 +1,441 @@
+    allowance_fetcher: F,
+}
+
+/// Chain ID for Mainnet.
+const MAINNET_CHAIN_ID: u64 = 1;
+
+impl MatchaSolver<GPv2Settlement> {
+    pub fn new(settlement_contract: GPv2Settlement, chain_id: u64) -> Result<Self> {
+        ensure!(
+            chain_id == MAINNET_CHAIN_ID,
+            "Matcha solver only supported on Mainnet",
+        );
+
+        let allowance_fetcher = settlement_contract.clone();
+        Ok(Self {
+            settlement_contract,
+            allowance_fetcher,
+            client: Box::new(DefaultMatchaApi::default()),
+        })
+    }
+}
+impl<F> MatchaSolver<F> {
+    fn web3(&self) -> DynWeb3 {
+        self.settlement_contract.raw_instance().web3()
+    }
+}
+
+impl<F> std::fmt::Debug for MatchaSolver<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("MatchaSolver")
+    }
+}
+
+#[async_trait::async_trait]
+impl<F: AllowanceFetching> SingleOrderSolving for MatchaSolver<F> {
+    async fn settle_order(&self, order: LimitOrder) -> Result<Option<Settlement>> {
+        let (swap, mut settlement) = match order.kind {
+            OrderKind::Sell => {
+                let query = SwapQuery {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: Some(order.sell_amount),
+                    buy_amount: None,
+                    slippage_percentage: Slippage::basis_points(STANDARD_MATCHA_SLIPPAGE_BPS)
+                        .unwrap(),
+                    skip_validation: Some(true),
+                };
+
+                tracing::debug!("querying Matcha swap api with {:?}", query);
+                let swap = self.client.get_swap(query).await?;
+                tracing::debug!("proposed Matcha swap is {:?}", swap);
+
+                if swap.buy_amount < order.buy_amount {
+                    tracing::debug!("Order limit price not respected");
+                    return Ok(None);
+                }
+                let settlement = Settlement::new(hashmap! {
+                    order.sell_token => swap.sell_amount,
+                    order.buy_token => swap.buy_amount,
+                });
+                (swap, settlement)
+            }
+            OrderKind::Buy => {
+                let query = SwapQuery {
+                    sell_token: order.sell_token,
+                    buy_token: order.buy_token,
+                    sell_amount: None,
+                    buy_amount: Some(order.buy_amount),
+                    slippage_percentage: Slippage::basis_points(STANDARD_MATCHA_SLIPPAGE_BPS)
+                        .unwrap(),
+                    // From the api documentation:
+                    // SlippagePercentage(Optional): The maximum acceptable slippage in % of the buyToken amount if sellAmount is provided, the maximum acceptable slippage in % of the sellAmount amount if buyAmount is provided. This parameter will change over time with market conditions.
```

Comment:
>This parameter will change over time with market conditions.
Not that I know what they mean, but I think their front-end queries this API with different slippage parameters depending on market conditions.

>"maximum acceptable slippage in % of the sellAmount" so it would adjust the sell amount?
Yes,  I was expecting the same! But I tested it and they don't do it. I will dig into it a little bit more on Monday

Reply:
Can you clean up this comment please? I think it's such an important consideration that it should go on top of the method (rather than in one of the branches). Maybe you can re-summarize your observations (potentially with links to the simulations which I found very useful). I personally find their documentation quote horrible to understand.